lib/mini_racer/truffleruby.rb

```diff
@@ -0,0 +1,355 @@
+        [
+          (x) => { return (x instanceof Object || x instanceof Array) && !(x instanceof Date) && !(x instanceof Function) },
+          (x) => { return x instanceof Date },
+          (x) => { return x.getTime(x) },
+          (x) => { return typeof x === 'symbol' },
+          (x) => { var r = x.description; return r === undefined ? 'undefined' : r },
+          (x) => { return new Date(x) },
+          (x) => { return new Array(x) },
+        ]
+      CODE
+    end
+
+    def dispose_unsafe
+      @context.close
+    end
+
+    def eval_unsafe(str, filename)
+      @entered = true
+      if !@has_entered && @snapshot
+        snapshot_src = encode(@snapshot.instance_variable_get(:@source))
+        begin
+          eval_in_context(snapshot_src, filename)
+        rescue Polyglot::ForeignException => e
+          raise RuntimeError, e.message, e.backtrace
+        end
+      end
+      @has_entered = true
+      raise RuntimeError, "TruffleRuby does not support eval after stop" if @stopped
+      raise TypeError, "wrong type argument #{str.class} (should be a string)" unless str.is_a?(String)
+      raise TypeError, "wrong type argument #{filename.class} (should be a string)" unless filename.nil? || filename.is_a?(String)
+
+      str = encode(str)
+      begin
+        translate do
+          eval_in_context(str, filename)
+        end
+      rescue Polyglot::ForeignException => e
+        raise RuntimeError, e.message, e.backtrace
+      rescue ::RuntimeError => e
+        if @current_exception
+          e = @current_exception
+          @current_exception = nil
+          raise e
+        else
+          raise e, e.message
+        end
+      end
+    ensure
+      @entered = false
+    end
+
+    def call_unsafe(function_name, *arguments)
+      @entered = true
+      if !@has_entered && @snapshot
+        src = encode(@snapshot.instance_variable_get(:source))
+        begin
+          eval_in_context(src)
+        rescue Polyglot::ForeignException => e
+          raise RuntimeError, e.message, e.backtrace
+        end
+      end
+      @has_entered = true
+      raise RuntimeError, "TruffleRuby does not support call after stop" if @stopped
+      begin
+        translate do
+          function = eval_in_context(function_name)
+          function.call(*convert_ruby_to_js(arguments))
+        end
+      rescue Polyglot::ForeignException => e
+        raise RuntimeError, e.message, e.backtrace
+      end
+    ensure
+      @entered = false
+    end
+
+    def create_isolate_value
+      # Returning a dummy object since TruffleRuby does not have a 1-1 concept with isolate.
+      # However, code and ASTs are shared between contexts.
+      Isolate.new
+    end
+
+    def isolate_mutex
+      @isolate_mutex
+    end
+
+    def translate
+      convert_js_to_ruby yield
+    rescue Object => e
+      message = e.message
+      if @current_exception
+        raise @current_exception
+      elsif e.message.start_with?('SyntaxError:')
```

Comment:
I'm seeing a bunch of errors in this line when I'm running specs from Discourse that are specific to `MiniRacer`.

```
bundle exec rspec spec/lib/pretty_text_spec.rb
```

Here's the error:
```
NoMethodError:
undefined method `start_with?' for nil:NilClass
# /home/gerhard/.gem/truffleruby/3.0.3/bundler/gems/mini_racer-33ab13dc022e/lib/mini_racer/truffleruby.rb:189:in `translate'
# /home/gerhard/.gem/truffleruby/3.0.3/bundler/gems/mini_racer-33ab13dc022e/lib/mini_racer/truffleruby.rb:131:in `eval_unsafe'
# /home/gerhard/.gem/truffleruby/3.0.3/bundler/gems/mini_racer-33ab13dc022e/lib/mini_racer.rb:228:in `block (2 levels) in eval'
# /home/gerhard/.gem/truffleruby/3.0.3/bundler/gems/mini_racer-33ab13dc022e/lib/mini_racer.rb:369:in `timeout'
# /home/gerhard/.gem/truffleruby/3.0.3/bundler/gems/mini_racer-33ab13dc022e/lib/mini_racer.rb:227:in `block in eval'
# /home/gerhard/.gem/truffleruby/3.0.3/bundler/gems/mini_racer-33ab13dc022e/lib/mini_racer.rb:225:in `synchronize'
# /home/gerhard/.gem/truffleruby/3.0.3/bundler/gems/mini_racer-33ab13dc022e/lib/mini_racer.rb:225:in `eval'
# ./lib/pretty_text.rb:245:in `block in markdown'
# ./lib/pretty_text.rb:605:in `block in protect'
# ./lib/pretty_text.rb:604:in `synchronize'
# ./lib/pretty_text.rb:604:in `protect'
# ./lib/pretty_text.rb:188:in `markdown'
# ./lib/pretty_text.rb:317:in `cook'
# ./spec/lib/pretty_text_spec.rb:1657:in `block (3 levels) in <top (required)>'
# ./spec/rails_helper.rb:278:in `block (2 levels) in <top (required)>'
# /home/gerhard/.gem/truffleruby/3.0.3/gems/webmock-3.14.0/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'
# <internal:core> core/kernel.rb:376:in `load'
# <internal:core> core/kernel.rb:376:in `load'
# <internal:core> core/kernel.rb:376:in `load'
# <internal:core> core/kernel.rb:376:in `load'
# <internal:core> core/kernel.rb:376:in `load'
# <internal:core> core/kernel.rb:376:in `load'
# ------------------
# --- Caused by: ---
# Polyglot::ForeignException:
```

Reply:
I think that's most likely caused by passing a Ruby String to JS, and JS not yet handling that correctly (calling Ruby methods on it vs JS String prototype functions).

@bjfish You need to handle String like this in `convert_ruby_to_js`:
```ruby
        when String
          Truffle::Interop.as_truffle_string value
```
from https://github.com/bjfish/mini_racer/pull/1#discussion_r916804519 and https://github.com/rails/execjs/pull/115/files

Also add the `if e.message && e.message.start_with?('SyntaxError:')` on this line to help debug.

Reply:
The handling of String is already there, but String needs to be removed from `when nil, true, false, Integer, Float, String`